### PR TITLE
Nn 2195 add appointment clashes

### PR DESCRIPTION
**Pull request by Robc24**
- added list of 'removed' clashing prisoners to view

### CODE REVIEW

File: backend/controllers/bulkAppointmentsFormRedirect.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+module.exports = (req, res) => {
+  // eslint-disable-next-line no-param-reassign
```
Comment (jsrobertson) [inline backend/controllers/bulkAppointmentsFormRedirect.js]: `// eslint-disable-next-line no-param-reassign` no longer required

File: backend/controllers/bulkAppointmentsClashes.js [hunk 2]
```diff
@@ -82,6 +82,17 @@ const bulkAppointmentsClashesFactory = (elite2Api, logError) => {
     } = req.session
 
     const remainingPrisoners = prisonersListed.filter(prisoner => !req.body[prisoner.offenderNo])
+    const removedPrisoners = prisonersListed.filter(prisoner => req.body[prisoner.offenderNo])
```
Comment (jsrobertson) [inline backend/controllers/bulkAppointmentsClashes.js]:
```suggestion
    const removedPrisoners = prisonersListed.filter(prisoner => req.body[prisoner.offenderNo]).map(prisoner => ({
      offenderNo: prisoner.offenderNo,
      lastName: prisoner.lastName,
    }))
```

Don't need the separate `prisonersRemoved` then

File: backend/controllers/bulkAppointmentsClashes.js [hunk 3]
```diff
@@ -82,6 +82,17 @@ const bulkAppointmentsClashesFactory = (elite2Api, logError) => {
...
+
+    const prisonersRemoved = removedPrisoners.map(prisoner => ({
+      offenderNo: prisoner.offenderNo,
+      lastName: prisoner.lastName,
+    }))
+
+    if (prisonersRemoved && prisonersRemoved.length) {
+      // eslint-disable-next-line no-param-reassign
```
Comment (jsrobertson) [inline backend/controllers/bulkAppointmentsClashes.js]: `// eslint-disable-next-line no-param-reassign` no longer required

File: backend/controllers/bulkAppointmentsClashes.js [hunk 4]
```diff
@@ -82,6 +82,17 @@ const bulkAppointmentsClashesFactory = (elite2Api, logError) => {
     } = req.session
 
     const remainingPrisoners = prisonersListed.filter(prisoner => !req.body[prisoner.offenderNo])
+    const removedPrisoners = prisonersListed.filter(prisoner => req.body[prisoner.offenderNo])
+
+    const prisonersRemoved = removedPrisoners.map(prisoner => ({
+      offenderNo: prisoner.offenderNo,
+      lastName: prisoner.lastName,
+    }))
```
Comment (jsrobertson) [inline backend/controllers/bulkAppointmentsClashes.js]:
```suggestion
```

File: backend/controllers/bulkAppointmentsClashes.js (hunk 2 above)
Comment (jsrobertson) [reply backend/controllers/bulkAppointmentsClashes.js]: or use a .reduce, probably better.

Comment (jsrobertson) [inline backend/controllers/bulkAppointmentsClashes.js]: Could probably use .reduce and go through `prisonersListed` once.

Comment (jsrobertson) [reply backend/controllers/bulkAppointmentsClashes.js]: I was thinking something like...
```
const [removedPrisoners, remainingPrisoners] = prisonersListed.reduce(
      (prisonersSeparated, prisoner) => {
        prisonersSeparated[req.body[prisoner.offenderNo] ? 0 : 1].push(prisoner)
        return prisonersSeparated
      },
      [[], []]
    )
```

Happy to discuss tomorrow though.

Comment (Robc24) [reply backend/controllers/bulkAppointmentsClashes.js]: Nice, I'd done this:

`    const prisonersRemoved = prisonersListed.reduce((removedList, prisoner) => {
      req.body[prisoner.offenderNo] && removedList.push({
        offenderNo: prisoner.offenderNo,
        lastName: prisoner.lastName
      });
      return removedList;
    }, [])`

But that caters for both nicely.

File: backend/controllers/bulkAppointmentsFormRedirect.js [hunk 5]
```diff
@@ -0,0 +1,5 @@
+module.exports = (req, res) => {
+  req.session.data = []
```
Comment (JamieDraperUK) [inline backend/controllers/bulkAppointmentsFormRedirect.js]: Be better setting this to null.

**Pull request merged**